### PR TITLE
Addition of Tooltip feature to Rendering settings

### PR DESCRIPTION
**Pull request by nasehim7**

### CODE REVIEW

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 1]
```diff
@@ -437,6 +437,7 @@ namespace
             create_direct_link("adaptive_sampler.enable_diagnostics",   "adaptive_pixel_renderer.enable_diagnostics");
 
             load_directly_linked_values(config);
+
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Remove blank line.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 2]
```diff
@@ -787,7 +820,6 @@ namespace
           : LightingEnginePanel("Unidirectional Path Tracer", parent)
         {
             fold();
-
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Put back blank line.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 3]
```diff
@@ -876,15 +911,21 @@ namespace
...
 
         void create_pt_advanced_max_ray_intensity_settings(QVBoxLayout* parent)
         {
             QDoubleSpinBox* max_ray_intensity = create_double_input("advanced.max_ray_intensity", 0.0, 1.0e9, 3, 0.1);
+            max_ray_intensity->setToolTip(m_params_metadata.get_path("pt.max_ray_intensity.help"));
+            
             QCheckBox* unlimited_ray_intensity = create_checkbox("advanced.unlimited_ray_intensity", "Unlimited");
             parent->addLayout(create_form_layout("Max Ray Intensity:", create_horizontal_group(max_ray_intensity, unlimited_ray_intensity)));
             connect(unlimited_ray_intensity, SIGNAL(toggled(bool)), max_ray_intensity, SLOT(setDisabled(bool)));
+
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Remove blank line.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 4]
```diff
@@ -998,6 +1042,7 @@ namespace
 
         void create_photon_tracing_settings(QVBoxLayout* parent)
         {
+
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Remove blank line.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 5]
```diff
@@ -1008,12 +1053,19 @@ namespace
...
+            sublayout->addRow("Light Photons:", light_photons);
+            sublayout->addRow("Environment Photons:", environment_photons);
+
+            light_photons->setToolTip(m_params_metadata.get_path("sppm.light_photons_per_pass.help"));
+            environment_photons->setToolTip(m_params_metadata.get_path("sppm.env_photons_per_pass.help"));
         }
 
         void create_radiance_estimation_settings(QVBoxLayout* parent)
         {
+
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Remove blank line.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 6]
```diff
@@ -501,17 +504,20 @@ namespace
...
             connect(
                 samples_spinbox, SIGNAL(valueChanged(const int)),
                 SLOT(slot_changed_uniform_sampler_samples(const int)));
 
-            m_uniform_sampler_decorrelate_pixels =
-                create_checkbox("uniform_sampler.decorrelate_pixels", "Decorrelate Pixels");
+            m_uniform_sampler_decorrelate_pixels = create_checkbox("uniform_sampler.decorrelate_pixels", "Decorrelate Pixels");
             layout->addWidget(m_uniform_sampler_decorrelate_pixels);
+            m_uniform_sampler_decorrelate_pixels->setToolTip(m_params_metadata.get_path("uniform_pixel_renderer.decorrelate_pixels.help"));
+
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Remove redundant blank line.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 7]
```diff
@@ -651,14 +666,25 @@ namespace
 
         void create_bounce_settings(QFormLayout* layout, const string& prefix)
         {
-            const string widget_base_key = prefix + ".bounces.";
 
+            const string widget_base_key = prefix + ".bounces.";
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Revert this change.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 8]
```diff
@@ -651,14 +666,25 @@ namespace
...
             QSpinBox* max_bounces = create_integer_input(widget_base_key + "max_bounces", 0, 10000, 1);
             QCheckBox* unlimited_bounces = create_checkbox(widget_base_key + "unlimited_bounces", "Unlimited");
+            
+            if(prefix == "photon_tracing")
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Insert a space after `if` (and all such keywords: `for`, `while`, `do`, `else`, etc.)

Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: That test doesn't look right actually.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 9]
```diff
@@ -651,14 +666,25 @@ namespace
...
+                max_bounces->setToolTip(m_params_metadata.get_path("sppm.photon_tracing_max_path_length.help"));
+            else
+                max_bounces->setToolTip(m_params_metadata.get_path("pt.max_path_length.help"));
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: That also doesn't look right. I think you should construct the path to the `help` component with the prefix, if possible, rather than testing for a few values of the prefix.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 10]
```diff
@@ -651,14 +666,25 @@ namespace
...
+
             layout->addRow("Max Bounces:", create_horizontal_group(max_bounces, unlimited_bounces));
             connect(unlimited_bounces, SIGNAL(toggled(bool)), max_bounces, SLOT(setDisabled(bool)));
 
-            layout->addRow("Russian Roulette Start Bounce:", create_integer_input(widget_base_key + "rr_start_bounce", 1, 10000, 1));
+            QSpinBox* russian_roulette_start = create_integer_input(widget_base_key + "rr_start_bounce", 1, 10000, 1);
+            russian_roulette_start->setToolTip(m_params_metadata.get_path("pt.rr_min_path_length.help"));
+
+            layout->addRow("Russian Roulette Start Bounce:", russian_roulette_start);
+
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Remove these two blank lines.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 11]
```diff
@@ -651,14 +666,25 @@ namespace
...
+            else
+                max_bounces->setToolTip(m_params_metadata.get_path("pt.max_path_length.help"));
+
             layout->addRow("Max Bounces:", create_horizontal_group(max_bounces, unlimited_bounces));
             connect(unlimited_bounces, SIGNAL(toggled(bool)), max_bounces, SLOT(setDisabled(bool)));
 
-            layout->addRow("Russian Roulette Start Bounce:", create_integer_input(widget_base_key + "rr_start_bounce", 1, 10000, 1));
+            QSpinBox* russian_roulette_start = create_integer_input(widget_base_key + "rr_start_bounce", 1, 10000, 1);
+            russian_roulette_start->setToolTip(m_params_metadata.get_path("pt.rr_min_path_length.help"));
+
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Remove this blank line.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 12]
```diff
@@ -700,8 +726,8 @@ namespace
         DistributionRayTracerPanel(const Configuration& config, QWidget* parent = 0)
           : LightingEnginePanel("Distribution Ray Tracer", parent)
         {
-            fold();
 
+            fold();
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Revert this change.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 13]
```diff
@@ -757,21 +783,28 @@ namespace
...
-            sublayout->addLayout(create_form_layout("Light Samples:", create_double_input("advanced.dl.light_samples", 0.0, 1000000.0, 3, 1.0)));
+            QDoubleSpinBox* light_samples = create_double_input("advanced.dl.light_samples", 0.0, 1000000.0, 3, 1.0);
+            sublayout->addLayout(create_form_layout("Light Samples:", light_samples));
+            light_samples->setToolTip(m_params_metadata.get_path("drt.dl_light_samples.help"));
         }
 
         void create_drt_advanced_ibl_settings(QVBoxLayout* parent)
         {
             QGroupBox* groupbox = new QGroupBox("Image-Based Lighting");
             parent->addWidget(groupbox);
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Remove this blank line to keep together things that go together.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 14]
```diff
@@ -862,7 +894,10 @@ namespace
             QHBoxLayout* sublayout = create_horizontal_layout();
             layout->addLayout(sublayout);
 
-            sublayout->addLayout(create_form_layout("Light Samples:", create_double_input("advanced.dl.light_samples", 0.0, 1000000.0, 3, 1.0)));
+            QDoubleSpinBox* light_samples = create_double_input("advanced.dl.light_samples", 0.0, 1000000.0, 3, 1.0);
+
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Remove blank line.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 15]
```diff
@@ -876,15 +911,21 @@ namespace
             QHBoxLayout* sublayout = create_horizontal_layout();
             layout->addLayout(sublayout);
 
-            sublayout->addLayout(create_form_layout("Environment Samples:", create_double_input("advanced.ibl.env_samples", 0.0, 1000000.0, 3, 1.0)));
+            QDoubleSpinBox* env_samples = create_double_input("advanced.ibl.env_samples", 0.0, 1000000.0, 3, 1.0);
+
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Remove blank line.

**Comment by nasehim7** [comment]
Update 1: Resolved the styling issues by placing the tooltip just after the creation of qt components so that tooltip creation gets standardized.
Update 2: Working on prefix check in create_bounce_settings()
